frame/executive/src/lib.rs

```diff
@@ -294,6 +294,11 @@ where
 		parent_hash: &System::Hash,
 		digest: &Digest,
 	) {
+		// Reset events before apply runtime upgrade hook.
+		// This is required to preserve events from runtime upgrade hook.
+		// This means the format of all the event related storages must always be compatible.
+		<frame_system::Pallet<System>>::reset_events();
```

Comment:
Please add a test to ensure this behavior. There should already be some test for runtime upgrade below, just send some event and check it is present.